### PR TITLE
Changes to Release pipeline to include GarnetServer nuget and other changes to the process

### PR DESCRIPTION
**Pull request by darrenge**
Sizable PR with the two main goals:
1) Getting GarnetServer released as separate nuget package 
2) Have only one place where we maintain a version number for our release.  
A secondary goal was to get away from .nuspec and use dotnet pack to create the nuget packages.

Directory.Build.props - this gives us the single place for the version as well as the shared meta data of the nuget packages

azure-pipelines-external-release.yml - uses dotnet pack; packs GarnetServer; no longer has the hard coded build name that needed to be updated

extract-version.ps1 - gets build name \ version from Directory.Build.props and sets version for yml and updates build number 

Garnet.host.csproj - csproj that gets built that provides all the files for the Microsoft.Garnet nuget package. Modified so dotnet pack can be used instead of nuget pack

GarnetServer.csproj - mostly just moving things out to the shared location of Directory.Build.props

Garnet.sln - deleted the Garnet.nuspec as it is no longer needed since using dotnet pack

This will fix issue #793

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -1,14 +1,40 @@
...
+	<!-- Versioning property for builds and packages -->
+	<PropertyGroup>
+		<VersionPrefix>1.0.40</VersionPrefix>
+	</PropertyGroup>
+
+	<PropertyGroup Label="Targets">
     <TargetFrameworks>net8.0</TargetFrameworks>
   </PropertyGroup>
-
+
```
Comment (badrishc) [inline Directory.Build.props:11]: nit: remove whitespace here

File: libs/host/Garnet.host.csproj [hunk 2]
```diff
@@ -29,4 +31,12 @@
...
\ No newline at end of file
+  <ItemGroup>
+    <!-- Include files from 8.0 root directory and runtimes in Nuget pacakge - don't want other folders in 8.0 root though  -->
+    <Content Include="..\host\bin\Release\net8.0\*" Pack="true" PackagePath="lib\net8.0" />
+    <Content Include="..\host\bin\Release\net8.0\runtimes\**\*" Pack="true" PackagePath="runtimes" />
+    <None Include="..\..\README.md" Pack="true" PackagePath="/"/>
+  </ItemGroup>
+
+</Project>
+
```
Comment (badrishc) [inline libs/host/Garnet.host.csproj:42]: nit: remove newline at end of file

File: libs/host/Garnet.host.csproj [hunk 3]
```diff
@@ -29,4 +31,12 @@
     <InternalsVisibleTo Include="Garnet.test" Key="0024000004800000940000000602000000240000525341310004000001000100011b1661238d3d3c76232193c8aa2de8c05b8930d6dfe8cd88797a8f5624fdf14a1643141f31da05c0f67961b0e3a64c7120001d2f8579f01ac788b0ff545790d44854abe02f42bfe36a056166a75c6a694db8c5b6609cff8a2dbb429855a1d9f79d4d8ec3e145c74bfdd903274b7344beea93eab86b422652f8dd8eecf530d2" />
   </ItemGroup>
 
-</Project>
\ No newline at end of file
+  <ItemGroup>
+    <!-- Include files from 8.0 root directory and runtimes in Nuget pacakge - don't want other folders in 8.0 root though  -->
+    <Content Include="..\host\bin\Release\net8.0\*" Pack="true" PackagePath="lib\net8.0" />
```
Comment (badrishc) [inline libs/host/Garnet.host.csproj:36]: do we need to include _everything_ in this path?

File: main/GarnetServer/GarnetServer.csproj [hunk 4]
```diff
@@ -3,28 +3,10 @@
...
-    <PackageRequireLicenseAcceptance>True</PackageRequireLicenseAcceptance>
-    <Description>
-      Garnet is a remote cache-store from Microsoft Research, that offers strong performance (throughput and latency),
-      scalability, storage, recovery, cluster sharding, key migration, and replication features. Garnet uses the Redis RESP wire
-      protocol and can work with existing Redis clients.
-    </Description>
-    <PackageReleaseNotes>See https://github.com/microsoft/garnet for details.</PackageReleaseNotes>
-    <Copyright>© Microsoft Corporation. All rights reserved.</Copyright>
-    <Tags>Garnet key-value store cache dictionary hashtable concurrent persistent remote cluster Redis RESP</Tags>
+    <PackageReadmeFile>readme.md</PackageReadmeFile>
```
Comment (badrishc) [inline main/GarnetServer/GarnetServer.csproj:9]: can we make this `README.md` (uppercase)

File: Garnet.sln [hunk 5]
```diff
@@ -42,7 +42,6 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution
 		Dockerfile.chiseled = Dockerfile.chiseled
 		Dockerfile.nanoserver = Dockerfile.nanoserver
 		Dockerfile.ubuntu = Dockerfile.ubuntu
-		Garnet.nuspec = Garnet.nuspec
```
Comment (badrishc) [inline Garnet.sln:45]: Add `extract_version.ps1` here?

File: .azure/pipelines/azure-pipelines-external-release.yml [hunk 6]
```diff
@@ -1,22 +1,20 @@
...
 ###################################### 
-name: 1.0.39
 trigger:
   branches:
     include:
     - main
   paths:
     include:
-    - .azure/pipelines/azure-pipelines-external-release.yml
+    - Directory.Build.props
```
Comment (badrishc) [inline .azure/pipelines/azure-pipelines-external-release.yml:11]: minor nit: remove whitespace at the end of sentence

File: .azure/pipelines/azure-pipelines-external-release.yml [hunk 7]
```diff
@@ -36,25 +42,17 @@ jobs:
...
     displayName: dotnet build
     enabled: True
     inputs:
-      projects: '**/Garnet.*.csproj'
+      projects: '**/Garnet*.csproj'
       arguments: -c Release
 
   - task: PowerShell@2
     displayName: 'Publish the GarnetServer binaries'
+    enabled: True
```
Comment (badrishc) [inline .azure/pipelines/azure-pipelines-external-release.yml:55]: why is this line addition necessary?

File: Directory.Build.props [hunk 8]
```diff
@@ -1,14 +1,40 @@
 <Project>
 
-  <PropertyGroup Label="Targets">
+	<!-- Versioning property for builds and packages -->
+	<PropertyGroup>
+		<VersionPrefix>1.0.40</VersionPrefix>
```
Comment (badrishc) [inline Directory.Build.props:5]: It seems we will run the release pipeline every time `Directory.Build.Props` changes which might be more frequent. This could potentially be fixed as follows:
1. Create a `Version.props` file that only has the version-related property group
```
<Project>
  <PropertyGroup>
    <VersionPrefix>1.0.40</VersionPrefix>
  </PropertyGroup>
</Project>
```
3. Import this file in `Directory.Build.props` as something like:
    `<Import Project="Version.props" />`
4. Change the YML trigger to `Version.props`

File: Directory.Build.props [hunk 9]
```diff
@@ -1,14 +1,40 @@
...
-
+	
   <PropertyGroup Label="Configuration">
     <LangVersion>latest</LangVersion>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
   </PropertyGroup>
 
+	<!-- Common properties for nuget packages -->
```
Comment (PaulusParssinen) [inline Directory.Build.props:19]: Great to see the move from `nuspec` to SDK-style properties 🎉

File: libs/host/Garnet.host.csproj (hunk 3 above)
Comment (darrenge) [reply libs/host/Garnet.host.csproj:36]: This was from Marc so I am not 100% sure but I believe so.

File: .azure/pipelines/azure-pipelines-external-release.yml (hunk 7 above)
Comment (darrenge) [reply .azure/pipelines/azure-pipelines-external-release.yml:55]: There are enabled: lines at each task which allows for easier development where you can disable all the tasks you want why developing pipelines.  I can remove them all.

**Review by badrishc — APPROVED** [review]

File: libs/host/Garnet.host.csproj (hunk 3 above)
Comment (darrenge) [reply libs/host/Garnet.host.csproj:36]: We get .pdb in library AND garnetserver nuget packages (as well as pdbs in the symbols package). This is what I was talking about offline. In earlier version of library we did NOT get the .pbs in Nuget package directly but we are now.

File: .azure/pipelines/azure-pipelines-external-release.yml (hunk 7 above)
Comment (badrishc) [reply .azure/pipelines/azure-pipelines-external-release.yml]: either keep all or remove all. it appeared from diff that only this section had it added, so.

Comment (darrenge) [reply .azure/pipelines/azure-pipelines-external-release.yml]: Removed all

File: Directory.Build.props (hunk 9 above)
Comment (darrenge) [reply Directory.Build.props:31]: We should have it out there soon. Just need to figure out how to use donet pack in conjunction with Directory.Build.props, GarnetServer.csproj Garnet.host.csproj and the release yml to NOT put .pdb in the garnet-server.nupkg and Microsoft.Garnet.nupkg.  We still want a .snupkg file that contains the symbols though.

Tried:
  <ItemGroup>
    <None Update="**\*.pdb">
      <Pack>false</Pack>
    </None>
  </ItemGroup>

and:
<AllowedOutputExtensionsInPackageBuildOutputFolder>dll</AllowedOutputExtensionsInPackageBuildOutputFolder>

Scratching my head over this one.

File: main/GarnetServer/GarnetServer.csproj [hunk 10]
```diff
@@ -4,12 +4,10 @@
     <OutputType>Exe</OutputType>
     <ServerGarbageCollection>true</ServerGarbageCollection>
     <PackageId>garnet-server</PackageId>
-    <PackAsTool>true</PackAsTool>
```
Comment (badrishc) [inline main/GarnetServer/GarnetServer.csproj:10]: why is this removed? isnt this what packs the project as a dotnet tool?

Comment (darrenge) [reply main/GarnetServer/GarnetServer.csproj:10]: Yes ... that is true but was trying to confirm something that I was seeing locally.

Comment (darrenge) [reply main/GarnetServer/GarnetServer.csproj:10]: Putting it back to PackAsTool

File: libs/host/Garnet.host.csproj (hunk 3 above)
Comment (darrenge) [reply libs/host/Garnet.host.csproj]: We decided to go with embedding the debug symbols in the DLLs, so there will not be any .snupkg files

**Review by badrishc — APPROVED** [review]

**Review by badrishc — APPROVED** [review]
